pupil_src/shared_modules/video_capture/utils.py

```diff
@@ -155,3 +165,144 @@ def check_slice(self, frame_gray):
+        return self._videos
+
+    @property
+    def containers(self) -> Sequence[Video]:
+        return [av.open(video.path) for video in self.videos]
+
+    @property
+    def lookup_loc(self) -> str:
+        return os.path.join(self.rec, f"{self.name}_lookup.npy")
+
+    def fetch_videos(self) -> Iterator[Video]:
+        yield from (
+            Video(loc)
+            for ext in self.video_exts
+            for loc in glob.iglob(
+                os.path.join(self.rec, f"{self.name}*.{ext}"))
+        )
+
+    def build_lookup(self):
+        """
+        The lookup table is a np.recarray containing entries
+        for each (virtual and real) frame.
+
+        Each entry consists of 3 values:
+            - container_idx: Corresponding self.videos index
+            - container_frame_idx: Frame index within the container
+            - timestamp: Recorded or virtual Pupil timestamp
+
+        container_idx entries of value -1 indicate a virtual frame.
+
+        The lookup table can be easiliy filtered for real frames:
+            lookup = lookup[lookup.container_idx > -1]
+
+        Use case:
+        Given a Pupil timestamp, one can use bisect to find the corresponding
+        lookup entry index. From there, one can lookup the corresponding
+        container, load it if necessary, and calculate the target PTS
+        """
+        loaded_ts = self._loaded_ts_sorted()
+        filled_ts = self._fill_gaps(loaded_ts)
+        lookup = self._setup_lookup(filled_ts)
+        lookup.timestamp = filled_ts
+        lookup.container_idx = -1  #
+        for container_idx, vid in enumerate(self.videos):
+            mask = np.isin(lookup.timestamp, vid.timestamps)
+            lookup.container_frame_idx[mask] = np.arange(vid.timestamps.size)
+            try:
+                vid.check_container()  # check if container is corrupt
+                lookup.container_idx[mask] = container_idx
+            except av.AVError:
+                logger.warning(f"{vid.name} is corrupt")
+        self.lookup = lookup
+
+    def save_lookup(self):
+        np.save(self.lookup_loc, self.lookup)
+
+    def load_lookup(self):
+        self.lookup = np.load(self.lookup_loc).view(np.recarray)
+
+    def _loaded_ts_sorted(self) -> np.ndarray:
+        loaded_ts = [vid.timestamps for vid in self.videos]
+        all_ts = np.concatenate(loaded_ts)
+        return all_ts
+
+    def _fill_gaps(self, timestamps: np.ndarray) -> np.ndarray:
+        time_diff = np.diff(timestamps)
+        median_time_diff = np.median(time_diff)
+        gap_start_idc = np.flatnonzero(time_diff > 5 * median_time_diff)
+        gap_stop_idc = gap_start_idc + 1
+        gap_fill_starts = timestamps[gap_start_idc] + median_time_diff
+        gap_fill_stops = timestamps[gap_stop_idc]
+
+        all_ts = [timestamps]
+        for start, stop in zip(gap_fill_starts, gap_fill_stops):
+            all_ts.append(np.arange(start, stop, median_time_diff))
```

Comment:
Should we use numpy's array instead of list for speed up?

Reply:
Not in this case, since appending numpy arrays is very slow in comparison to adding entries to python lists. `np.concatenate` appends all single numpy arrays at once in the end.